core/src/bundle_stage.rs

```diff
@@ -301,80 +301,103 @@ impl BundleStage {
+                bundle_receiver,
+                decision_maker,
+                consumer,
+                unprocessed_bundle_storage,
+                &mut last_metrics_update,
                 &mut bundle_stage_metrics,
                 &mut bundle_stage_leader_metrics,
             ) {
-                Ok(_) | Err(RecvTimeoutError::Timeout) => (),
-                Err(RecvTimeoutError::Disconnected) => break,
+                error!("Bundle stage error: {e:?}");
+                break;
             }
+        }
+    }
 
-            let bundle_storage = unprocessed_bundle_storage.bundle_storage().unwrap();
-            bundle_stage_metrics.increment_current_buffered_bundles_count(
-                bundle_storage.unprocessed_bundles_len() as u64,
-            );
-            bundle_stage_metrics.increment_current_buffered_packets_count(
-                bundle_storage.unprocessed_packets_len() as u64,
-            );
-            bundle_stage_metrics.increment_cost_model_buffered_bundles_count(
-                bundle_storage.cost_model_buffered_bundles_len() as u64,
-            );
-            bundle_stage_metrics.increment_cost_model_buffered_packets_count(
-                bundle_storage.cost_model_buffered_packets_len() as u64,
+    fn process_bundles(
+        bundle_receiver: &mut BundleReceiver,
+        decision_maker: &DecisionMaker,
+        consumer: &mut BundleConsumer,
+        unprocessed_bundle_storage: &mut UnprocessedTransactionStorage,
+        last_metrics_update: &mut Instant,
+        bundle_stage_metrics: &mut BundleStageLoopMetrics,
+        bundle_stage_leader_metrics: &mut BundleStageLeaderMetrics,
+    ) -> Result<(), RecvTimeoutError> {
+        if !unprocessed_bundle_storage.is_empty()
+            || last_metrics_update.elapsed() >= SLOT_BOUNDARY_CHECK_PERIOD
+        {
+            let (_, process_buffered_packets_time) = measure!(
+                Self::process_buffered_bundles(
+                    decision_maker,
+                    consumer,
+                    unprocessed_bundle_storage,
+                    bundle_stage_leader_metrics,
+                ),
+                "process_buffered_packets",
             );
-            bundle_stage_metrics.maybe_report(1_000);
+            bundle_stage_leader_metrics
+                .leader_slot_metrics_tracker()
+                .increment_process_buffered_packets_us(process_buffered_packets_time.as_us());
+            *last_metrics_update = Instant::now();
+        }
+
+        if let Err(RecvTimeoutError::Disconnected) = bundle_receiver.receive_and_buffer_bundles(
+            unprocessed_bundle_storage,
+            bundle_stage_metrics,
+            bundle_stage_leader_metrics,
+        ) {
+            return Err(RecvTimeoutError::Disconnected);
         }
+
+        let bundle_storage = unprocessed_bundle_storage.bundle_storage().unwrap();
+        bundle_stage_metrics.increment_current_buffered_bundles_count(
```

Comment:
inherited a bug with these metrics? seems like should be setters rather than incrementing. cc: @buffalu

Reply:
updated